File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums); //배열 정렬(오름차순)
        List<List<Integer>> list=new ArrayList();
        for(int i1=0;i1<nums.length-2;i1++){
            if(i1>0 && nums[i1]==nums[i1-1])
                continue;
            int i2=i1+1;
            int i3=nums.length-1;
            while(i2<i3){
                int sum=nums[i1]+nums[i2]+nums[i3];
                if(sum==0){
                    List<Integer> tmp=new ArrayList();
                    tmp.add(nums[i1]);
                    tmp.add(nums[i2]);
                    tmp.add(nums[i3]);
                    list.add(tmp);
                    i3--;
                    while(i2<i3 && nums[i3]==nums[i3+1]){ //중복
                        i3--;  
                    }
                }
                else if(sum>0){
                    i3--;
                }else{ //sum이 0보다 작은경우 
                    i2++;
                }
            }
        }
        return list;
    }
}

File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        list = []
        if len(nums)>2:
            for i1 in range(len(nums)-2):
                if i1>0 and nums[i1]==nums[i1-1]:
                    continue
                i2 = i1+1
                i3 = len(nums)-1
                while(i2<i3):
                    sum = nums[i1]+nums[i2]+nums[i3]
                    if sum==0:
                        list.append([nums[i1],nums[i2],nums[i3]])
                        i3 -= 1
                        while (i2<i3 and nums[i3]==nums[i3+1]):
                            i3 -= 1
                    elif sum>0:
                        i3 -= 1
                    else:
                        i2 += 1
        return list